### PR TITLE
Update Handling of SSO Linking

### LINKED ISSUE #900: ActiveRecord::RecordInvalid: Validation failed: Email has already been taken
View details in Rollbar: [https://app.rollbar.com/a/ualbertalib/fix/item/dmp_assistant/534](https://app.rollbar.com/a/ualbertalib/fix/item/dmp_assistant/534)

```

ActiveRecord::RecordInvalid: Validation failed: Email has already been taken
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/validations.rb", line 80, in raise_validation_error
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/validations.rb", line 53, in save!
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/transactions.rb", line 302, in block in save!
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/transactions.rb", line 354, in block in with_transaction_returning_status
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/connection_adapters/abstract/database_statements.rb", line 320, in block in transaction
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/connection_adapters/abstract/transaction.rb", line 319, in block in within_new_transaction
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/concurrency/load_interlock_aware_monitor.rb", line 26, in block (2 levels) in synchronize
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/concurrency/load_interlock_aware_monitor.rb", line 25, in handle_interrupt
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/concurrency/load_interlock_aware_monitor.rb", line 25, in block in synchronize
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/concurrency/load_interlock_aware_monitor.rb", line 21, in handle_interrupt
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/concurrency/load_interlock_aware_monitor.rb", line 21, in synchronize
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/connection_adapters/abstract/transaction.rb", line 317, in within_new_transaction
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/connection_adapters/abstract/database_statements.rb", line 320, in transaction
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/transactions.rb", line 350, in with_transaction_returning_status
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/transactions.rb", line 302, in save!
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/suppressor.rb", line 48, in save!
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/persistence.rb", line 55, in create!
  File "/var/www/sites/dmp/app/models/user.rb", line 193, in create_from_provider_data
  File "/var/www/sites/dmp/app/controllers/users/omniauth_callbacks_controller.rb", line 29, in openid_connect
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/action_controller/metal/basic_implicit_render.rb", line 6, in send_action
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/abstract_controller/base.rb", line 228, in process_action
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/action_controller/metal/rendering.rb", line 30, in process_action
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/abstract_controller/callbacks.rb", line 42, in block in process_action
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/callbacks.rb", line 117, in block in run_callbacks
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/turbo-rails-2.0.5/lib/turbo-rails.rb", line 24, in with_request_id
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/turbo-rails-2.0.5/app/controllers/concerns/turbo/request_id_tracking.rb", line 10, in turbo_tracking_request_id
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/callbacks.rb", line 126, in block in run_callbacks
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actiontext-6.1.7.8/lib/action_text/rendering.rb", line 20, in with_renderer
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actiontext-6.1.7.8/lib/action_text/engine.rb", line 59, in block (4 levels) in <class:Engine>
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/callbacks.rb", line 126, in instance_exec
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/callbacks.rb", line 126, in block in run_callbacks
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/callbacks.rb", line 137, in run_callbacks
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/abstract_controller/callbacks.rb", line 41, in process_action
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/action_controller/metal/rescue.rb", line 22, in process_action
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/action_controller/metal/instrumentation.rb", line 34, in block in process_action
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/notifications.rb", line 203, in block in instrument
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/notifications/instrumenter.rb", line 24, in instrument
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activesupport-6.1.7.8/lib/active_support/notifications.rb", line 203, in instrument
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/action_controller/metal/instrumentation.rb", line 33, in process_action
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/action_controller/metal/params_wrapper.rb", line 249, in process_action
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/activerecord-6.1.7.8/lib/active_record/railties/controller_runtime.rb", line 27, in process_action
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/abstract_controller/base.rb", line 165, in process
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionview-6.1.7.8/lib/action_view/rendering.rb", line 39, in process
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/action_controller/metal.rb", line 190, in dispatch
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/action_controller/metal.rb", line 254, in dispatch
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/action_dispatch/routing/route_set.rb", line 50, in dispatch
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/action_dispatch/routing/route_set.rb", line 33, in serve
  File "/var/www/sites/dmp/vendor/ruby/3.1.0/gems/actionpack-6.1.7.8/lib/action_dispa

### PR DESCRIPTION
**Pull request by aaronskiba**
Fixes #900
Changes proposed in this PR:
- Update the handling of SSO linking in some user case scenarios.

### CODE REVIEW

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -22,11 +22,19 @@ def openid_connect
...
-          # Register and sign in
+      if current_user.nil? # if user is not signed in (They clicked the SSO sign in button)
+        if user.nil? # If an entry does not exist in the identifiers table for the chosen SSO account
           user = User.create_from_provider_data(auth)
+          if user.nil? # if a user was NOT created (a match was found for User.find_by(email: auth.info.email)
+            # Do not link SSO credentials for the signed out, existing user
+            flash[:alert] = _('The email you selected has not yet been linked to an existing account.<br>' \
+                              "Please sign in via the 'Sign in' button and navigate to the " \
+                              "'Edit Profile' section of DMP Assistant.<br>" \
+                              'From there you can link an email and enable single sign on access.<br>')
```
Comment (lagoan) [inline app/controllers/users/omniauth_callbacks_controller.rb]: We need to make this message simpler. Something along the lines of:

_The email provided has already been registered as an existing account._ Please sign in via the 'Sign in' button and navigate to the 'Edit Profile' section of DMP Assistant. From there you can link an email and enable single sign on access

Youll notice I only modified the first sentence. Maybe in can still be further improved ?

File: app/models/user.rb [hunk 2]
```diff
@@ -188,7 +188,7 @@ def self.from_omniauth(auth)
   def self.create_from_provider_data(provider_data)
     user = User.find_by email: provider_data.info.email
 
-    return user if user
+    return if user
```
Comment (lagoan) [inline app/models/user.rb]: I would like to see the changes made here combined with the proposed modifications in PR #902 to make sure it handles both cases.

**Review by lagoan — CHANGES_REQUESTED** [review]
This is a good start.

There are a couple of places that could be improved.

File: app/controllers/users/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (aaronskiba) [reply app/controllers/users/omniauth_callbacks_controller.rb]: I agree, it is a little complex. Maybe we should get Marcus to decide on the final result?

File: app/models/user.rb (hunk 2 above)
Comment (aaronskiba) [reply app/models/user.rb]: I've added a new commit to include that proposed idea: https://github.com/portagenetwork/roadmap/pull/907/commits/1985f1e8e0352ad9e485a4f7a6fd6c3a68b7ebcf

File: app/controllers/users/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (aaronskiba) [reply app/controllers/users/omniauth_callbacks_controller.rb]: How about the following?

```
flash[:alert] = _('You must be signed in to link the provided email.<br>' \
                              "Please sign in via the 'Sign in' button and navigate to the " \
                              "'Edit Profile' section of DMP Assistant.<br>" \
                              'From there you can link the provided email to enable single sign on access.<br>')
```

Screenshot from 2024-09-17 09-08-49

Comment (lagoan) [reply app/controllers/users/omniauth_callbacks_controller.rb]: We should be clear that the email is already used as a system account. The user needs to sign in using that system account with that email, and then they can link the account.

**Comment by closenma** [comment]
Proposed English:
`That email appears to be associated with an existing account. \
Sign into your existing account, and you can link with SSO from the 'Edit Profile' page.`
Translation is standing by to get us a quick version in French

**Comment by closenma** [comment]
Got back from translation - small rephrase on both - use these versions, please

English: `That email appears to be associated with an existing account. \ Sign into your existing account, and you can link that account with SSO from the 'Edit Profile' page.`

French: `Il paraît que cette adresse couriel est dejà associée à un compte existant. \ Veuillez vous connecter à votre compte existant, et vous pouvez le lier à l'aide de la fonction SSO à partir de la page 'Modifier le profil'.`

**Review by lagoan — DISMISSED** [review]
LGTM!

**Review by lagoan — DISMISSED** [review]
LGTM!

**Review by lagoan — APPROVED** [review]

**Comment by aaronskiba** [comment]
`bundle exec rspec ./spec/integration/openid_connect_sso_spec.rb:42` has an outdated button name. The test passes when I update it. If I change it I lose the PR approval, but I have no one to re-approve. Merging as is.
```
$ bundle exec rspec ./spec/integration/openid_connect_sso_spec.rb:42
To use retry middleware with Faraday v2.0+, install `faraday-retry` gem
Copying Bootstrap glyphicons to the public directory ...
Copying TinyMCE skins to the public directory ...
Run options: include {:locations=>{"./spec/integration/openid_connect_sso_spec.rb"=>[42]}}

Randomized with seed 11395
.

Top 1 slowest examples (1.91 seconds, 55.2% of total time):
  Openid_connection SSO with correct credentials does not create SSO link when user is signed out and SSO email is an existing account email
    1.91 seconds ./spec/integration/openid_connect_sso_spec.rb:42

Finished in 3.47 seconds (files took 4.85 seconds to load)
1 example, 0 failures

Randomized with seed 11395
```

**Pull request merged**